feat: add `lapack/base/dlarfg` (#7109)
stdlib-js/stdlib: lib/node_modules/@stdlib/lapack/base/dlarfg/README.md

You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# dlarfg

> LAPACK routine to generate a real elementary reflector `H` of order `N` such that applying `H` to a vector `[alpha; x]` zeros out `X`.

<section class="usage">

## Usage

```javascript
var dlarfg = require( '@stdlib/lapack/base/dlarfg' );
```

#### dlarfg( N, X, incx, out )

Generates a real elementary reflector `H` of order `N` such that applying `H` to a vector `[alpha; x]` zeros out `X`.

```javascript
var Float64Array = require( '@stdlib/array/float64' );

var X = new Float64Array( [ 2.0, 3.0, 4.0 ] );
var out = new Float64Array( [ 4.0, 0.0 ] );

dlarfg( 4, X, 1, out );
// X => <Float64Array>[ ~0.19, ~0.28, ~0.37 ]
// out => <Float64Array>[ ~-6.7, ~1.6 ]
```

The function has the following parameters:

-   **N**: number of rows/columns of the elementary reflector `H`.

> -   **X**: overwritten by the vector `V` on exit, expects `N - 1` indexed elements [`Float64Array`][mdn-float64array].

Reviewer: ```suggestion
-   **X**: a [`Float64Array`][mdn-float64array] which is overwritten by the vector `V`. Should have `N - 1` indexed elements.
```

Where is `V` defined?

-   **incx**: stride length of `X`.

> -   **out**: output [`Float64Array`][mdn-float64array], the first element of `out` represents alpha and the second element of `out` represents `tau`.

Reviewer: ```suggestion
-   **out**: output [`Float64Array`][mdn-float64array]. The first element of `out` represents alpha and the second element of `out` represents `tau`.
```

-   **strideOut**: stride length of `out`.
-   **offsetOut**: starting index of `out`.

While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying buffer, the offset parameters support indexing semantics based on starting indices. For example,

```javascript
var Float64Array = require( '@stdlib/array/float64' );

var X = new Float64Array( [ 0.0, 2.0, 3.0, 4.0 ] );
var out = new Float64Array( [ 0.0, 4.0, 0.0 ] );

dlarfg.ndarray( 4, X, 1, 1, out, 1, 1 );
// X => <Float64Array>[ 0.0, ~0.19, ~0.28, ~0.37 ]
// out => <Float64Array>[ 0.0, ~6.7, ~1.6 ]
```

</section>

<!-- /.usage -->

<section class="notes">

## Notes

-   `H` is a Householder matrix with the form `H = I - tau * [1; v] * [1, v^T]`, where `tau` is a scalar and `v` is a vector.
-   the input vector is `[alpha; x]`, where `alpha` is a scalar and `X` is a real `(n-1)`-element vector.
-   the result of applying `H` to `[alpha; x]` is `[beta; 0]`, with `beta` being a scalar and the rest of the vector zeroed.
-   if all elements of `X` are zero, then `tau = 0` and `H` is the identity matrix.

> -   otherwise, `1 <= tau <= 2`

Reviewer: I suggest moving this up to an introduction section where you can use TeX equations.
